File: 112.cpp
struct Solution {
    bool dfs(TreeNode *node, int res){
        if (node == nullptr){
            return false;
        }
        res -= node->val;
        if (node->left == nullptr && node->right == nullptr){
            return res == 0;
        }
        return this->dfs(node->left, res) || this->dfs(node->right, res);
    }
    inline bool hasPathSum(TreeNode* root, int targetSum) {
        return this->dfs(root, targetSum);
    }
};

File: sol/112.py
class Solution:
    def hasPathSum(self, root: Optional[TreeNode], targetSum: int) -> bool:
        def dfs(node, curr):
            if node is None:
                return False
            curr -= node.val
            if node.left is None and node.right is None:
                return curr == 0
            return dfs(node.left, curr) or dfs(node.right, curr)
        return dfs(root, targetSum)
